src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs

```diff
@@ -0,0 +1,306 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipelines.Tests
+{
+    public class PipeReaderStreamTests
+    {
+        public delegate Task<int> ReadAsyncDelegate(Stream stream, byte[] data);
+
+        [Theory]
+        [MemberData(nameof(ReadCalls))]
+        public async Task ReadingFromPipeReaderStreamReadsFromUnderlyingPipeReader(ReadAsyncDelegate readAsync)
+        {
+            byte[] helloBytes = Encoding.ASCII.GetBytes("Hello World");
+            var pipe = new Pipe();
+            await pipe.Writer.WriteAsync(helloBytes);
+            pipe.Writer.Complete();
+
+            var stream = new PipeReaderStream(pipe.Reader);
+
+            var buffer = new byte[1024];
+            int read = await readAsync(stream, buffer);
+
+            Assert.Equal(helloBytes, buffer.AsSpan(0, read).ToArray());
+            pipe.Reader.Complete();
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadCalls))]
+        public async Task AsStreamReturnsPipeReaderStream(ReadAsyncDelegate readAsync)
+        {
+            byte[] helloBytes = Encoding.ASCII.GetBytes("Hello World");
+            var pipe = new Pipe();
+            await pipe.Writer.WriteAsync(helloBytes);
+            pipe.Writer.Complete();
+
+            Stream stream = pipe.Reader.AsStream();
+
+            var buffer = new byte[1024];
+            int read = await readAsync(stream, buffer);
+
+            Assert.Equal(helloBytes, buffer.AsSpan(0, read).ToArray());
+            pipe.Reader.Complete();
+        }
+
+        [Fact]
+        public async Task ReadingWithSmallerBufferWorks()
+        {
+            byte[] helloBytes = Encoding.ASCII.GetBytes("Hello World");
+            var pipe = new Pipe();
+            await pipe.Writer.WriteAsync(helloBytes);
+            pipe.Writer.Complete();
+
+            Stream stream = pipe.Reader.AsStream();
+
+            var buffer = new byte[5];
+            int read = await stream.ReadAsync(buffer);
+
+            Assert.Equal(5, read);
+            Assert.Equal(helloBytes.AsSpan(0, 5).ToArray(), buffer.AsSpan(0, read).ToArray());
+
+            // Verify that the buffer is partially consumed and we can read the rest from the PipeReader directly
+            ReadResult result = await pipe.Reader.ReadAsync();
+            Assert.Equal(helloBytes.AsSpan(5).ToArray(), result.Buffer.ToArray());
+            pipe.Reader.AdvanceTo(result.Buffer.End);
+
+            pipe.Reader.Complete();
+        }
+
+        [Fact]
+        public async Task EndOfPipeReaderReturnsZeroBytesFromReadAsync()
+        {
+            var pipe = new Pipe();
+            pipe.Writer.Complete();
+
+            Stream stream = pipe.Reader.AsStream();
+
+            var buffer = new byte[5];
+            var read = await stream.ReadAsync(buffer);
```

Comment:
Read twice here. We had a bug with it working only once.

Reply:
Read what twice? From the Stream?

Reply:
Yes, past the end of the stream twice.

Reply:
Ah, I'll write a different test. I want to verify that I can interleave reading from the stream and pipe.